### PR TITLE
Datafusion aggregate

### LINKED ISSUE #464: [DF] Port Aggregate logic to Datafusion
**Is your feature request related to a problem? Please describe.**
Update all existing "Aggregate" logic to replace Apache Calcite data structures with Apache Datafusion data structures.

**Describe the solution you'd like**
Replace instances of Calcite [LogicalAggregate](https://calcite.apache.org/javadocAggregate/org/apache/calcite/rel/logical/LogicalAggregate.html) with correlating instances of Datafusion [Aggregate](https://docs.rs/datafusion/latest/datafusion/logical_plan/plan/struct.Aggregate.html). This includes any base classes, functions, members that might also rely on that logic.

**Describe alternatives you've considered**
None

**Additional context**
Success criteria is for all pytests in test_groupby.py to be passing

### LINKED ISSUE #472: [DF] Add GitHub workflow for cargo fmt and cargo clippy
**Is your feature request related to a problem? Please describe.**
It is standard practice to have CI run `cargo fmt --check` and `cargo clippy` on pull requests on Rust projects. It makes it easier for other developers to contribute if we are all using the same code formatting.

**Describe the solution you'd like**
Set up GtiHub actions to run `cargo fmt` and `cargo clippy` checks. We can look at the [DataFusion Rust workflow](https://github.com/apache/arrow-datafusion/blob/master/.github/workflows/rust.yml) as an example.

**Describe alternatives you've considered**
None

**Additional context**
None

### LINKED ISSUE #478: [DF] Build against master branch of DataFusion
**Is your feature request related to a problem? Please describe.**
The master branch of DataFusion has bug fixes and new features that we want to use from dask-sql so we may want to build against the master branch in the short term rather than depending on version 7.0.0.

**Describe the solution you'd like**
Change cargo.toml dependency:

``` toml
datafusion = { git="https://github.com/apache/arrow-datafusion/" }
```

**Describe alternatives you've considered**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by jdye64**
Adds the support for Aggregate and Group By in queries. 

This closes #464 
This closes #472
This closes #478

### CODE REVIEW

**Comment by charlesbluca** [comment]
Is there any way we can get CI running for this currently so that the uncommented tests run?

**Comment by jdye64** [comment]
Yep. I knew it wouldn't work on first pass but will work it out here in a few commits coming soon.

**Comment by jdye64** [comment]
@andygrove would really like your input now that the Rust Lint and Clippy steps have been added. I am certainly not an expert there so any input you have would be appreciated.

File: .github/workflows/release.yml
Comment (charlesbluca) [inline .github/workflows/release.yml:27]: We won't need to worry about this until it's time to merge into `main`, but probably important to check that we can pip install the outputs of

```
python setup.py sdist bdist_wheel
```

After running through the env setup of this script

File: .github/workflows/rust.yml
Comment (charlesbluca) [inline .github/workflows/rust.yml:1]: Is there a reason these checks need a separate workflow and can't be handled in the [pre-commit workflow](https://github.com/dask-contrib/dask-sql/blob/main/.github/workflows/style.yml)?

File: .github/workflows/test.yml
Comment (charlesbluca) [inline .github/workflows/test.yml:48]: Is this step still necessary? It doesn't look like we're caching the output of the dask_planner build so we could just do this in each test job?

File: .github/workflows/test-upstream.yml
Comment (charlesbluca) [inline .github/workflows/test-upstream.yml:30]: We can probably delete this entire caching step now if we aren't caching any build output

File: README.md
Comment (charlesbluca) [inline README.md:117]: No recompilation needs to happen on changes to the Rust code?

**Comment by jdye64** [comment]
Had a thought here. I currently just have the pytests that do not work yet commented out. I feel like I should change that to have something like `@pytest.mark.skip(reason="Incomplete Datafusion logic")` what do others think?

File: continuous_integration/recipe/meta.yaml
Comment (charlesbluca) [inline continuous_integration/recipe/meta.yaml:43]: If we need to add pyarrow to the conda package, I'll note that @rjzamora commented that our min version is quite low here (0.15.1 is over 2 years old now) - I'm tempted to bump to 1.0.0 since that's what's installed in for our independent cluster testing but I'll defer to him since he probably knows more than me here

**Review by charlesbluca — COMMENTED** [review]
Just a quick look through some of the meta stuff - will dig into the Python changes later

Also I notice that `dask_planner` has an empty README - do we want to add a separate README there giving some documentation on the Rust code, or do we think it's okay just removing (not sure if it needs to be there for some reason)?

File: continuous_integration/recipe/meta.yaml
Comment (jdye64) [reply continuous_integration/recipe/meta.yaml:43]: I would like to bump the version as well. After you ask which version let me know and I will change it. Our use is very simple for pyarrow, we just use it to get the Arrow Types since that is what Datafusion uses so it makes the types mapping more straight forward. I saw that to point out we likely don't have to worry about breaking changes from newer versions as those types almost never change.

File: README.md
Comment (jdye64) [reply README.md:117]: They would just run another `python ./setup.py install` guess we should add that in the README though.

File: .github/workflows/test.yml
Comment (jdye64) [reply .github/workflows/test.yml:48]: I'm not the best with github workflows. While the maven step is not necessary a step that caches the Rust compilation would be

File: .github/workflows/rust.yml
Comment (jdye64) [reply .github/workflows/rust.yml:1]: Not that I can think of. I'll move them to pre-commit

File: continuous_integration/recipe/meta.yaml
Comment (ayushdg) [reply continuous_integration/recipe/meta.yaml:43]: After updating it's also worth confirming it doesn't conflict with the pyarrow requirements that cuDF has and doesn't overwrite the pyarrow cuda builds that cuDF requests

Comment (jdye64) [reply continuous_integration/recipe/meta.yaml:43]: Good point

Comment (jdye64) [reply continuous_integration/recipe/meta.yaml]: bumped to `7.0.0` which seems to be what was present in the [cudf repo](https://github.com/rapidsai/cudf/blob/branch-22.06/conda/environments/cudf_dev_cuda11.5.yml#L21)

File: .github/workflows/test-upstream.yml
Comment (ayushdg) [reply .github/workflows/test-upstream.yml:30]: Do we still need to include a build step above for this workflow to run?

File: .github/workflows/test.yml
Comment (andygrove) [inline .github/workflows/test.yml:55]: pedantic nit:

```suggestion
      - name: Build the Rust DataFusion bindings
```

File: README.md
Comment (andygrove) [inline README.md:169]: DataFusion isn't a top-level Apache project but part of the Apache Arrow project, so we should probably refer to it either as just "DataFusiuon" or by the full name "Apache Arrow DataFusion".

```suggestion
For the first step, Apache Arrow DataFusion needs to know about the columns and types of the dask dataframes, therefore some Rust code to store this information for dask dataframes are defined in `dask_planner`.
```

File: dask_planner/src/sql/logical.rs
Comment (andygrove) [inline dask_planner/src/sql/logical.rs:93]:
```suggestion
                "Unable to compute DaskTable from DataFusion LogicalPlan",
```

File: dask_sql/physical/rex/core/function.py
Comment (andygrove) [inline dask_sql/physical/rex/core/function.py:39]:
```suggestion
                "Failed to determine DataFusion Type in scalar_function.py"
```

File: .github/workflows/test.yml
Comment (andygrove) [inline .github/workflows/test.yml:3]:
```suggestion
# 1. build dask_planner (Apache Arrow DataFusion Rust bindings) on ubuntu
```

File: .github/workflows/test.yml
Comment (charlesbluca) [reply .github/workflows/test.yml:48]: Sure, happy to sync on this and figure out what files we would want to store as artifacts across the run

File: .github/workflows/test-upstream.yml
Comment (charlesbluca) [reply .github/workflows/test-upstream.yml:30]: Yeah, we would need to remove the `needs: build` line for this job since the `build` job has been removed, although based on conversation [here](https://github.com/dask-contrib/dask-sql/pull/471/files/9936c92930019e11e0e5e5dc7d3831c6ae1ca9e4#r849618927) I think we might prefer keeping the `build` jobs in and just altering them to build / cache the Rust code instead

**Review by andygrove — COMMENTED** [review]
The GitHub actions for Rust look good. I see the clippy output in the actions that ran. There is no output from cargo fmt but it appears to have run. I tested out the commands locally as well.

I left a bunch of pedantic suggestions about using `DataFusion` rather than `Datafusion` and also about using the term `Apache Arrow DataFusion` rather than `Apache DataFusion` since this isn't a top-level Apache project.

File: .github/workflows/test.yml
Comment (jdye64) [reply .github/workflows/test.yml:55]: Good catch. I'll do a global find/replace as I'm sure I did this in several places.

**Comment by jdye64** [comment]
How do people feel about leaving the old `planner` `.gitignore` rules in here for awhile? Seems like to me someone might still have them on their local development environment and might still accidentally commit them and we should leave them. I am open to suggestions however.

File: .github/workflows/rust.yml
Comment (jdye64) [reply .github/workflows/rust.yml]: @charlesbluca given my last comment what do you think? Should I move the logic to style.yaml or leave it in rust.yaml?

File: .github/workflows/test.yml
Comment (jdye64) [reply .github/workflows/test.yml]: removed this.

File: tests/integration/test_groupby.py
Comment (ayushdg) [inline tests/integration/test_groupby.py:403]: Same

**Review by ayushdg — CHANGES_REQUESTED** [review]
Thanks a lot for this huge effort of kicking off the port to datafusion. 
Just had a few questions but things generally lgtm. 

I noticed that gpuCI isn't running on this PR. @charlesbluca do we need anything on the gpuci side to enable this?

File: .github/docker-compose.yaml
Comment (charlesbluca) [reply .github/docker-compose.yaml:15]: Agree that we might want to use Dask as a guideline here vs. cuDF, as it's possible that we could end up with a pinning that's incompatible with upstream Dask.

Looking through Dask's CI, it looks like the lowest version being tested right now is 4.0.0, so I think that setting something like `pyarrow>=4.0.0` throughout seems like a reasonable choice

File: .github/workflows/test-upstream.yml
Comment (charlesbluca) [reply .github/workflows/test-upstream.yml:10]:
```suggestion
```

File: continuous_integration/environment-3.10-dev.yaml
Comment (charlesbluca) [reply continuous_integration/environment-3.10-dev.yaml:26]: As I said above think we could switch this to `pyarrow>=4.0.0`, which is what Dask is testing against

**Review by charlesbluca — COMMENTED** [review]
Thanks for the gpuCI catch @ayushdg, opened up https://github.com/rapidsai/gpuci-scripts/pull/810 to add `datafusion-sql-planner` to the list of branches gpuCI targets

File: dask_sql/physical/rex/core/function.py
Comment (jdye64) [reply dask_sql/physical/rex/core/function.py:39]: Yeah this is bad ... I think this deserves its own [issue](https://github.com/dask-contrib/dask-sql/issues/479) and to be covered in another PR. Is that ok @ayushdg?

File: dask_sql/physical/rex/core/call.py
Comment (jdye64) [reply dask_sql/physical/rex/core/call.py:726]: Yes, the test coverage for that operation can be found [here](https://github.com/dask-contrib/dask-sql/blob/f00d49836cc91e8b9df5d64496ad423ec8cc5e4e/tests/unit/test_call.py#L186)

File: tests/integration/test_compatibility.py
Comment (jdye64) [reply tests/integration/test_compatibility.py:283]: The name of the PR is quite misleading. While its name is "Datafusion aggregate" in reality it is the first PR for baseline removing all Java logic and doesn't necessarily mean that all `aggregate` functionality will work along with it yet. These tests have pieces that still will not work until subsequent PRs.

File: tests/integration/test_groupby.py
Comment (jdye64) [reply tests/integration/test_groupby.py:39]: The name of the PR is quite misleading. While its name is "Datafusion aggregate" in reality it is the first PR for baseline removing all Java logic and doesn't necessarily mean that all `aggregate` functionality will work along with it yet. These tests have pieces that still will not work until subsequent PRs.

File: tests/integration/test_groupby.py
Comment (jdye64) [reply tests/integration/test_groupby.py:403]: The name of the PR is quite misleading. While its name is "Datafusion aggregate" in reality it is the first PR for baseline removing all Java logic and doesn't necessarily mean that all `aggregate` functionality will work along with it yet. These tests have pieces that still will not work until subsequent PRs.

File: dask_sql/physical/rex/core/function.py
Comment (ayushdg) [reply dask_sql/physical/rex/core/function.py:39]: Thanks for raising the issue for tracking.This can be handled in a followup pr.

File: .github/workflows/rust.yml
Comment (charlesbluca) [inline .github/workflows/rust.yml:23]: IIUC we can remove this because it's being handled as part of the pre-commit checks now, correct?

File: .github/workflows/rust.yml
Comment (charlesbluca) [inline .github/workflows/rust.yml:53]: It looks like [pre-commit-rust](https://github.com/doublify/pre-commit-rust) has a hook for this?

https://github.com/doublify/pre-commit-rust/blob/eeee35a89e69d5772bdee97db1a6a898467b686e/hooks.yaml#L15-L21

Let me know if that works for this; my preference would be to consolidate the Rust linting into the pre-commit hooks so that we're doing the full linting on every commit.

File: .pre-commit-config.yaml
Comment (charlesbluca) [inline .pre-commit-config.yaml:20]: pre-commit throws up a warning whenever it runs because we aren't using a tagged version here - would you mind switching to

```suggestion
    rev: v1.0
```

File: .pre-commit-config.yaml
Comment (charlesbluca) [inline .pre-commit-config.yaml:25]: Think we should be able to get `clippy` running here with

```suggestion
        args: ['--manifest-path', './dask_planner/Cargo.toml', '--verbose', '--']
    -   id: cargo-clippy
        args: ['--manifest-path', './dask_planner/Cargo.toml', '--verbose', '--']
```

File: continuous_integration/recipe/meta.yaml
Comment (charlesbluca) [inline continuous_integration/recipe/meta.yaml:42]: Out of interest, what happens if we don't have `setuptools-rust` installed along with the conda package? Interested in why we need it as a runtime and host dependency

File: continuous_integration/recipe/meta.yaml
Comment (charlesbluca) [inline continuous_integration/recipe/meta.yaml:43]: I wouldn't add this to the conda recipe just yet - I'd ideally like to have more discussion around what package we want to consider "core" to the dask-sql conda package first (see https://github.com/dask-contrib/dask-sql/issues/435).

For now, I think just bumping the min version in the:

- docker compose files
- CI environments
- setup.py

Should be sufficient.

**Review by charlesbluca — CHANGES_REQUESTED** [review]

File: dask_sql/physical/rel/custom/alter.py
Comment (jdye64) [reply dask_sql/physical/rel/custom/alter.py:4]: Gotcha and yes that is primarily the reason so I will make that change

File: continuous_integration/recipe/meta.yaml
Comment (jdye64) [reply continuous_integration/recipe/meta.yaml:42]: Since the `setup.py` file has references to `Rust` extensions the build will fail if setuptools-rust is not present.

Comment (jdye64) [reply continuous_integration/recipe/meta.yaml]: Oh, I was mistaken and thought this was another file. Yes, this can be removed from this section and is in my latest commit. Thanks for catching that and I will resolve this.

**Comment by jdye64** [comment]
@charlesbluca how does one find the link for the gpuci runs?

**Comment by charlesbluca** [comment]
rerun tests

**Comment by charlesbluca** [comment]
rerun tests

**Comment by charlesbluca** [comment]
rerun tests

**Review by ayushdg — APPROVED** [review]
Updates lgtm! We should be good to merge on my end.

Last thing I wanted to discuss if there's scope to release the Python GIL in situations where we aren't directly working with python objects such as parsing the sql string and converting the ast into a datafusion plan. If it is, it might be worth raising an issue and following up in a separate pr.

File: .pre-commit-config.yaml
Comment (charlesbluca) [inline .pre-commit-config.yaml:13]: Any reason why we're bumping isort's pre-commit hook?

If we decide to keep this, we should also bump the version used in the conda environments and setup.py

Comment (charlesbluca) [inline .pre-commit-config.yaml]: Any reason why we're bumping isort's pre-commit hook?

If we decide to keep this, we should also bump the version used in the conda environments and setup.py

File: pre-commit.sh
Comment (charlesbluca) [inline pre-commit.sh:1]: Is this file necessary as part of the pre-commit hooks in `.pre-commit-config.yaml`?

Comment (charlesbluca) [inline pre-commit.sh:1]: Is this file necessary as part of the pre-commit hooks in `.pre-commit-config.yaml`?

**Review by charlesbluca — APPROVED** [review]

**Pull request merged**
